lib/journaled/transaction_ext.rb

```diff
@@ -0,0 +1,31 @@
+require 'active_record/connection_adapters/abstract/transaction'
+
+module Journaled
+  module TransactionExt
+    def initialize(*, **)
+      super.tap do
+        raise TransactionSafetyError, <<~MSG unless instance_variable_defined?(:@run_commit_callbacks)
+          Journaled::TransactionExt expects @run_commit_callbacks to be defined on Transaction!
+          This is an internal API that may have changed in a recent Rails release.
+          If you were not expecting to see this error, please file an issue here:
+          https://github.com/Betterment/journaled/issues
+        MSG
+      end
+    end
+
+    def before_commit_records
+      super.tap do
+        Writer.enqueue!(*_journaled_staged_events) if @run_commit_callbacks
```

Comment:
This mimics the behavior that the transaction records follow -- basically, if the transaction is configured to run callbacks, we'll write out the events here, otherwise we'll merge the events onto the parent transaction's staged events array (down on line 23).

This allows us to emit at most 1 job per actual COMMIT statement, even when we're using nested transactions / save states internally.